src/tracedAtomic.ml

```diff
@@ -109,167 +125,298 @@ let handler current_process_id runner =
-                        |> IntSet.of_seq in
-  CCVector.clear processes;
-  atomics_counter := 1;
-  match last_element init_schedule with
-  | (run_proc, run_op, run_ptr) ->
-    { procs; enabled = current_enabled; run_proc; run_op; run_ptr; backtrack = IntSet.empty }
-
-let rec explore func state clock last_access =
-  let s = last_element state in
-  List.iter (fun proc ->
-      let j = proc.proc_id in
-      let i = Option.bind proc.obj_ptr (fun ptr -> IntMap.find_opt ptr last_access) |> Option.value ~default:0 in
-      if i != 0 then begin
-        let pre_s = List.nth state (i-1) in
-        if IntSet.mem j pre_s.enabled then
-          pre_s.backtrack <- IntSet.add j pre_s.backtrack
-        else
-          pre_s.backtrack <- IntSet.union pre_s.backtrack pre_s.enabled
+  let procs =
+    IdMap.mapi (fun i p -> { proc_id = i; op = p.next_op; obj_ptr = p.next_repr }) !processes
+  in
+  let current_enabled =
+    IdMap.to_seq !processes
+    |> Seq.filter (fun (_, p) -> not p.finished)
+    |> Seq.map (fun (id, _) -> id)
+    |> IdSet.of_seq
+  in
+  let last_run = List.hd init_schedule in
+  { procs; enabled = current_enabled; run = last_run; backtrack = IdMap.empty }
+
+type category =
+  | Ignore
+  | Read
+  | Write
+  | Read_write
+
+let categorize = function
+  | Spawn | Start | Make -> Ignore
+  | Get -> Read
+  | Set -> Write
+  | Exchange | FetchAndAdd -> Read_write
+  | CompareAndSwap res ->
+      begin match !res with
+      | Unknown -> failwith "CAS unknown outcome" (* should not happen *)
+      | Success -> Read_write
+      | Failed -> Read_write
+      end
+
+let rec list_findi predicate lst i = match lst with
+  | [] -> None
+  | x::_ when predicate i x -> Some (i, x)
+  | _::xs -> list_findi predicate xs (i + 1)
+
+let list_findi predicate lst = list_findi predicate lst 0
+
+let mark_backtrack proc time state (last_read, last_write) =
+  let j = proc.proc_id in
+  let find ptr map = match IdMap.find_opt ptr map with
+    | None -> None
+    | Some lst ->
+        List.find_opt (fun (_, proc_id) -> proc_id <> j) lst
+  in
+  let find_loc proc =
+    match categorize proc.op, proc.obj_ptr with
+    | Ignore, _ -> None
+    | Read, Some ptr -> find ptr last_write
+    | (Write | Read_write), Some ptr -> max (find ptr last_read) (find ptr last_write)
+    | _ -> assert false
+  in
+  let rec find_replay_trace ~lower ~upper proc_id =
+    let pre_s = List.nth state (time - upper + 1) in
+    let replay_steps = List.filteri (fun k s -> k >= lower && k <= time - upper && s.run.proc_id = proc_id) state in
+    let replay_steps = List.rev_map (fun s -> s.run) replay_steps in
+    let causal p = match find_loc p with
+      | None -> true
+      | Some (k, _) -> k <= upper in
+    if List.for_all causal replay_steps
+    then if IdSet.mem proc_id pre_s.enabled
+         then Some replay_steps
+         else let is_parent k s = k > lower && k < time - upper && s.run.op = Spawn && s.run.obj_ptr = Some proc_id in
```

Comment:
Is the `k > lower && k < time - upper` condition needed? That is, isn't a `Spawn` of some `proc_id` unique within a single execution (state)?

Reply:
Yes the check on the interval is a defensive measure, we can't always backtrack to before our spawn... This whole function is really hard to read though :/